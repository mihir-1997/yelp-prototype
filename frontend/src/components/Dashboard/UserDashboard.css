.dashboard-wrapper {
    margin: 2% 0;
}

.restaurant-search-bar {
    padding-bottom: 2%;
    border-bottom: 1px lightgray solid;
}

.below-restaurant-search-bar {
    margin: 2% 2% 0;
}

.restaurant-search .form-group {
    padding: 0px;
}

/* .restaurant-search {
    position: fixed;
} */

.searchOptions,
.searchrestaurant{
    height: 50px
}

.searchsubmit {
    height: 50px;
    width: 50px;
    background-color: #f43939;
}

.filters-wrapper {
    padding: 3% 20px 0 50px;
    min-height: 80vh;
    border-right: 1px lightgray solid;
}

.filters-heading{
    border-bottom: 1px lightgray solid;
}

.filters-heading h5 {
    width: 100%;
    text-align: center;
}

.suggested-heading {
    margin: 20px 0 0 20px;
}

.delivery_filters {
    margin: 10px 0 0 20px;
}

.restaurants-wrapper {
    padding: 2% 2% 0 2%;
}

.restaurants {
    margin-top: 20px;
}

.maps-wrapper {
    max-height: 80vh;
}