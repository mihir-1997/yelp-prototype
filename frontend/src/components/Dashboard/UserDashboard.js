import React, { Component } from 'react'
import axios from 'axios';

import './UserDashboard.css'
import Restaurant from './Restaurant'
import Maps from './Maps'

class UserDashboard extends Component {

    constructor( props ) {
        super( props )
        this.state = {
            search: "",
            restautants: [],
            filtered_restaurants: [],
            latLongs: [],
            filtered_latLongs: [],
            filtered: "",
            selectedOption: "",
            curbside_pickup: false,
            dine_in: false,
            delivery: false
        }
    }

    onChange = ( item ) => {
        this.setState( {
            [ item.target.name ]: item.target.value
        } )
    }

    onClick = ( data, avgRatings ) => {
        return this.props.history.push( {
            pathname: '/restaurant',
            state: {
                id: data,
                avgRatings: avgRatings
            }
        } )
    }

    searchRestaurants = ( item ) => {
        item.preventDefault()
        axios.defaults.withCredentials = true;

        if ( this.state.search && this.state.selectedOption ) {
            axios.get( "http://localhost:3001/getrestaurantbysearch/" + this.state.selectedOption + "/" + this.state.search )
                .then( ( res ) => {
                    if ( res.status === 200 ) {
                        let ids = []
                        this.setState( {
                            filtered_restaurants: this.state.filtered_restaurants.filter( ( restaurant ) => {
                                if ( res.data.includes( restaurant.id ) ) {
                                    ids.push( restaurant.id )
                                    return true
                                }
                                return false
                            } ),
                            filtered_latLongs: this.state.latLongs.filter( latlong => ids.includes( latlong.id ) )
                        } )
                    }
                } )
                .catch( ( err ) => {
                    if ( err.response ) {
                        if ( err.response.status === 404 ) {
                            this.setState( {
                                filtered_restaurants: [],
                                filtered_latLongs: []
                            } )
                        } else if ( err.response.status === 500 ) {
                            console.log( err.response.message )
                        }
                    }
                } )
        } else {
            this.setState( {
                filtered_restaurants: this.state.restautants,
                filtered_latLongs: this.state.latLongs
            } )
        }
    }

    filterRestaurant = ( item ) => {
        if ( this.state[ this.state.filtered ] ) {
            let ids = []
            this.setState( {
                filtered_restaurants: this.state.filtered_restaurants.filter( ( restaurant ) => {
                    if ( restaurant[ this.state.filtered ] === 1 ) {
                        ids.push( restaurant.id )
                        return true
                    }
                    return false
                } ),
                filtered_latLongs: this.state.latLongs.filter( latlong => ids.includes( latlong.id ) )
            } )
        } else {
            this.setState( {
                filtered_restaurants: this.state.restautants,
                filtered: "",
                filtered_latLongs: this.state.latLongs
            } )
        }
    }

    onChangeFilter = ( item ) => {
        this.setState( {
            [ item.target.value ]: !this.state[ item.target.value ],
            filtered: item.target.value
        }, this.filterRestaurant )
    }

    componentDidMount () {
        axios.defaults.withCredentials = true;
        let id = localStorage.getItem( "id" )
        if ( id ) {
            axios.get( "http://localhost:3001/getAllRestaurants" )
                .then( ( res ) => {
                    if ( res.status === 200 ) {
                        this.setState( {
                            restautants: res.data.restaurants,
                            filtered_restaurants: res.data.restaurants,
                            latLongs: res.data.latlongs,
                            filtered_latLongs: res.data.latlongs
                        } )
                    }
                } )
                .catch( ( err ) => {
                    if ( err.response ) {
                        if ( err.response.status === 404 ) {
                            console.log( err.response.message )
                        } else if ( err.response.status === 500 ) {
                            console.log( err.response.message )
                        }
                    }
                } )
        } else {
            console.log( "No Id found in local storage" )
        }
    }

    render () {
        if ( localStorage.getItem( "active" ) !== "user" ) {
            this.props.history.goBack()
        }
        return (
            <div className="dashboard-wrapper">
                <div className="row restaurant-search-bar">
                    <div className="col-2"></div>
                    <div className="col-8 restaurant-search">
                        <form>
                            <div className="form-row">
                                <div className="form-group col-md-2">
                                    <select name="selectedOption" id="searchOptions" className="form-control searchOptions" onChange={ this.onChange }>
                                        <option name="searchByName" value="">Search By...</option>
                                        <option name="searchByName" value="dish">Dish</option>
                                        <option name="searchByName" value="cuisine">Cuisine</option>
                                        <option name="searchByName" value="location">Location</option>
                                        {/* <option name="searchByName" value="deliverymode">Mode of Delivery</option> */ }
                                    </select>
                                </div>
                                <div className="form-group col-md-8">
                                    <input type="text" className="form-control searchrestaurant" name="search" value={ this.state.search } onChange={ this.onChange } />
                                </div>
                                <div className="form-group col-md-2">
                                    <button type="button" className="btn searchsubmit" onClick={ this.searchRestaurants }>
                                        <svg xmlns="http://www.w3.org/2000/svg" className="icon icon-tabler icon-tabler-search" width="25" height="25" viewBox="0 0 24 24" strokeWidth="1.5" stroke="#ffffff" fill="none" strokeLinecap="round" strokeLinejoin="round">
                                            <path stroke="none" d="M0 0h24v24H0z" fill="none" />
                                            <circle cx="10" cy="10" r="7" />
                                            <line x1="21" y1="21" x2="15" y2="15" />
                                        </svg>
                                    </button>
                                </div>
                            </div>
                        </form>
                    </div>
                    <div className="col-2"></div>
                </div>
                {/* <div className="below-restaurant-search-bar"> */ }
                <div className="row">
                    <div className="col-2 filters-wrapper">
                        <div className="row filters-heading">
                            <h5>Filters</h5>
                        </div>
                        <div className="row">
                            <div className="col">
                                <div className="row suggested-heading">
                                    Suggested
                                </div>
                                <div className="row">
                                    <div className="delivery_filters">
                                        <input type="checkbox" name="delivery_method" value="curbside_pickup" onChange={ this.onChangeFilter } disabled={ this.state.dine_in || this.state.delivery } /> Curbside Pickup<br />
                                        <input type="checkbox" name="delivery_method" value="dine_in" onChange={ this.onChangeFilter } disabled={ this.state.curbside_pickup || this.state.delivery } /> Open Now<br />
                                        <input type="checkbox" name="delivery_method" value="delivery" onChange={ this.onChangeFilter } disabled={ this.state.curbside_pickup || this.state.dine_in } /> Yelp Delivery<br />
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div className="col-7 restaurants-wrapper">
                        <h2>Restaurants</h2>
                        <div className="row restaurants">
                            { this.state.filtered_restaurants.length ?
                                this.state.filtered_restaurants.map( ( restaurant, index ) => {
                                    return <Restaurant restautantData={ restaurant } index={ index } onClick={ this.onClick } key={ index } />
                                } )
                                : "No restaurants found" }
                        </div>
                    </div>
                    <div className="col-3 maps-wrapper">
                        <Maps latlongs={ this.state.filtered_latLongs } />
                    </div>
                </div>
            </div>
            // </div>
        )
    }
}
export default UserDashboard;
