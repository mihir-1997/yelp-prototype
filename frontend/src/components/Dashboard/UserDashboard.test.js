import React from 'react';
import { shallow } from 'enzyme';

import UserDashboard from './UserDashboard'

describe( 'UserDashboard', () => {

    let axios = require( "axios" );
    let MockAdapter = require( "axios-mock-adapter" );
    let mock = new MockAdapter( axios );

    const BACKEND_URL = process.env.REACT_APP_BACKEND_URL || "http://localhost"
    const BACKEND_PORT = process.env.REACT_APP_BACKEND_PORT || 3001

    const restaurants = [
        {
            address: "43 S 1st St",
            city: "San Jose",
            curbside_pickup: 1,
            delivery: 0,
            description: "We serve quality food!",
            dine_in: 1,
            email: "res@gmail.com",
            id: 1,
            image: "RestaurantImages/subway-1600419949316.jpeg",
            name: "Subway",
            phone_no: "(123)-456-7893",
            state: "CA",
            timings: "9:00 AM-11:00 PM",
            zipcode: 95113
        },
        {
            address: "1398 W San Carlos St",
            city: "San Jose",
            curbside_pickup: 0,
            delivery: 1,
            description: "Tasty Food!",
            dine_in: 1,
            email: "mcd@gmail.com",
            id: 4,
            image: "RestaurantImages/erik-mclea-1601630103191.jpg",
            name: "McDonalds",
            phone_no: "(123)-456-7892",
            state: "CA",
            timings: "9:00 AM-11:00 PM",
            zipcode: 95126
        }
    ]

    const filtered_restaurants = [
        {
            address: "43 S 1st St",
            city: "San Jose",
            curbside_pickup: 1,
            delivery: 0,
            description: "We serve quality food!",
            dine_in: 1,
            email: "res@gmail.com",
            id: 1,
            image: "RestaurantImages/subway-1600419949316.jpeg",
            name: "Subway",
            phone_no: "(123)-456-7893",
            state: "CA",
            timings: "9:00 AM-11:00 PM",
            zipcode: 95113
        }
    ]

    mock.onGet( BACKEND_URL + ":" + BACKEND_PORT + "/getAllRestaurants" ).reply( 200,
        { restaurants: restaurants, latlongs: [] }
    );

    mock.onGet( BACKEND_URL + ":" + BACKEND_PORT + "/getrestaurantbysearch/cuisine/indian" ).reply( 200,
        [ 1 ]
    );

    let component

    beforeEach( () => {
        localStorage.setItem( "active", "user" )
        localStorage.setItem( "id", 1 )
        component = shallow( <UserDashboard debug /> );
    } );

    it( 'should render correctly in "debug" mode', () => {
        component.instance()
        expect( component ).toMatchSnapshot();
    } );

    it( 'validate localstorage has values', () => {
        expect( localStorage.getItem( "id" ) ).toEqual( 1 )
        expect( localStorage.getItem( "active" ) ).toEqual( "user" )
    } );

    it( 'render after setting localstorage', async ( done ) => {
        component
            .instance()
            .componentDidMount()
            .then( () => {
                expect( component.state().restaurants ).toEqual( restaurants )
                expect( component.state().filtered_restaurants ).toEqual( restaurants )
                done()
            } )
        expect( component ).toMatchSnapshot();
    } );

    it( 'simulate checkbox to filter restaurants by curbside_pickup', async ( done ) => {
        component
            .instance()
            .componentDidMount()
            .then( () => {
                let restaurants_filter = component.find( 'input[type="checkbox"]' ).at( 0 )
                expect( component.state().filtered_restaurants ).toEqual( restaurants )
                restaurants_filter.simulate( 'change', { target: { checked: true, value: "curbside_pickup" } } )
                expect( component.state().filtered ).toEqual( "curbside_pickup" )
                expect( component.state().filtered_restaurants ).toEqual( filtered_restaurants )
                restaurants_filter.simulate( 'change', { target: { checked: false, value: "curbside_pickup" } } )
                expect( component.state().filtered ).toEqual( "" )
                expect( component.state().filtered_restaurants ).toEqual( restaurants )
                done()
            } )

        expect( component ).toMatchSnapshot();
    } )

    it( 'simulate search restaurants to search indian cuisine', async ( done ) => {
        component
            .instance()
            .componentDidMount()
            .then( () => {
                let search_options = component.find( 'select.searchOptions' )
                let search = component.find( 'input[type="text"]' )
                expect( component.state().filtered_restaurants ).toEqual( restaurants )
                search_options.simulate( 'keydown', { keyCode: 40 } )
                search_options.simulate( 'keydown', { keyCode: 13 } )
                search_options.simulate( 'change', { target: { name: "selectedOption", value: "cuisine" } } )
                expect( component.state().selectedOption ).toEqual( "cuisine" )
                search.simulate( 'change', { target: { name: "search", value: "indian" } } )
                expect( component.state().search ).toEqual( "indian" )
                component
                    .instance()
                    .searchRestaurants( {
                        preventDefault: () => {
                        }
                    } )
                    .then( () => {
                        expect( component.state().filtered_restaurants ).toEqual( filtered_restaurants )
                        done()
                    } )

                expect( component ).toMatchSnapshot();
            } )
    } )

    it( 'validate localstorage is empty', () => {
        localStorage.removeItem( "id" )
        localStorage.removeItem( "active" )
        expect( localStorage.store ).toEqual( {} )
    } );

} )