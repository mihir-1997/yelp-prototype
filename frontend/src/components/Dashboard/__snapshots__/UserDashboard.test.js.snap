// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDashboard render after setting localstorage 1`] = `ShallowWrapper {}`;

exports[`UserDashboard should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`UserDashboard simulate checkbox to filter restaurants by curbside_pickup 1`] = `ShallowWrapper {}`;

exports[`UserDashboard simulate search restaurants to search indian cuisine 1`] = `ShallowWrapper {}`;
